dv/uvm/core_ibex/sim.py

```diff
@@ -1,301 +1,512 @@
+                               '(used in string: {!r}).'
+                               .format(needle, string))
+        return string
+
+    return string.replace(needle, replacement)
+
+
+def subst_env_vars(string, env_vars):
+    '''Substitute environment variables in string
+
+    env_vars should be a string with a comma-separated list of environment
+    variables to substitute. For each environment variable, V, in the list, any
+    occurrence of <V> in string will be replaced by the value of the
+    environment variable with that name. If <V> occurs in the string but $V is
+    not set in the environment, an error is raised.
+
+    '''
+    env_vars = env_vars.strip()
+    if not env_vars:
+        return string
+
+    for env_var in env_vars.split(','):
+        env_var = env_var.strip()
+        needle = '<{}>'.format(env_var)
+        if needle in string:
+            value = os.environ.get(env_var)
+            if value is None:
+                raise RuntimeError('Cannot substitute {} in command because '
+                                   'the environment variable ${} is not set.'
+                                   .format(needle, env_var))
+            string = string.replace(needle, value)
+
+    return string
+
+
+def subst_cmd(cmd, enable_dict, opts_dict, env_vars):
+    '''Substitute options and environment variables in cmd
+
+    enable_dict should be a dict mapping names to bools. For each key, N, in
+    enable_dict, if enable_dict[N] is False, then all occurrences of <N> in cmd
+    will be replaced with ''. If enable_dict[N] is True, all occurrences of <N>
+    in cmd will be replaced with opts_dict[N].
+
+    If N is not a key in opts_dict, this is no problem unless cmd contains
+    <N>, in which case we throw a RuntimeError.
+
+    Finally, the environment variables are substituted as described in
+    subst_env_vars and any newlines are stripped out.
+
+    '''
+    for name, enable in enable_dict.items():
+        cmd = subst_opt(cmd, name, enable, opts_dict.get(name))
+
+    return subst_env_vars(cmd, env_vars).replace('\n', ' ')
+
+
+def subst_vars(string, var_dict):
+    '''Apply substitutions in var_dict to string
+
+    If var_dict[K] = V, then <K> will be replaced with V in string.'''
+    for key, value in var_dict.items():
+        string = string.replace('<{}>'.format(key), value)
+    return string
+
+
+def get_yaml_for_simulator(simulator, yaml_path):
+    '''Read yaml at yaml_path and find entry for simulator'''
+    logging.info("Processing simulator setup file : %s" % yaml_path)
+    for entry in read_yaml(yaml_path):
+        if entry.get('tool') == simulator:
+            return entry
+
+    raise RuntimeError("Cannot find RTL simulator {}".format(simulator))
+
+
+def get_simulator_cmd(simulator, yaml_path, enables):
+    '''Get compile and run commands for the testbench
+
+    simulator is the name of the simulator to use. yaml_path is the path to a
```

Comment:
Just a more nitty comment: could you retain a similar formatting of the function argument/return value docstrings to how it was before, with a colon mapping each argument to its description?

Reply:
Re the comment formatting, is this an internal Google style? I don't think I've seen it before. We can do that if you think it's important, but it will make the comments longer and I don't think they become any more informative. Since tidying up the patch set properly would take half an hour's work or so, I'm going to hold off on this for now: if you think this needs doing, let me know and I'll update the patches.

Reply:
Yep, the comment formatting is as per the open-source Google Python coding style guide, in particular [this section](http://google.github.io/styleguide/pyguide.html#383-functions-and-methods).
This function docstring formatting isn't the most important thing, as you already are providing some explanations of the arguments and so on, so I'm ok if you just leave it as is for this PR, and we can just leave this for some time in the future.